torchgeo/datasets/splits.py

```diff
@@ -0,0 +1,359 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""Dataset splitting utilities."""
+
+from copy import deepcopy
+from itertools import accumulate
+from math import floor
+from typing import List, Optional, Sequence, Tuple, Union
+
+from rtree.index import Index, Property
+from torch import Generator, default_generator, randint, randperm
+
+from ..datasets import GeoDataset
+from .utils import BoundingBox
+
+__all__ = (
+    "random_bbox_assignment",
+    "random_bbox_splitting",
+    "random_grid_cell_assignment",
+    "roi_split",
+    "time_series_split",
+)
+
+
+def _fractions_to_lengths(fractions: Sequence[float], total: int) -> Sequence[int]:
+    """Utility to divide a number into a list of integers according to fractions.
+
+    Args:
+        fractions: list of fractions
+        total: total to be divided
+
+    Returns:
+        List of lengths.
+
+    .. versionadded:: 0.5
+    """
+    lengths = [floor(frac * total) for frac in fractions]
+    remainder = int(total - sum(lengths))
+    # add 1 to all the lengths in round-robin fashion until the remainder is 0
+    for i in range(remainder):
+        idx_to_add_at = i % len(lengths)
+        lengths[idx_to_add_at] += 1
+    return lengths
+
+
+def random_bbox_assignment(
+    dataset: GeoDataset,
+    lengths: Sequence[Union[int, float]],
+    generator: Optional[Generator] = default_generator,
+) -> List[GeoDataset]:
+    """Split a GeoDataset randomly assigning its index's BoundingBoxes.
+
+    Args:
+        dataset: dataset to be split
+        lengths: lengths or fractions of splits to be produced
+        generator: (optional) generator used for the random permutation
+
+    Returns
+        A list of the subset datasets.
+
+    .. versionadded:: 0.5
+    """
+    if not (sum(lengths) == 1 or sum(lengths) == len(dataset)):
+        raise ValueError(
+            "Sum of input lengths must equal 1 or the length of dataset's index."
+        )
+
+    if any(n <= 0 for n in lengths):
+        raise ValueError("All items in input lengths must be greater than 0.")
+
+    if sum(lengths) == 1:
+        lengths = _fractions_to_lengths(lengths, len(dataset))
+
+    hits = list(dataset.index.intersection(dataset.index.bounds, objects=True))
+
+    hits = [
+        hits[i]
+        for i in randperm(sum(lengths), generator=generator)  # type: ignore[arg-type]
+    ]
+
+    new_indexes = [
+        Index(interleaved=False, properties=Property(dimension=3)) for _ in lengths
+    ]
+
+    for i, length in enumerate(lengths):
+        for j in range(length):  # type: ignore[arg-type]
+            hit = hits.pop()
+            new_indexes[i].insert(j, hit.bounds, hit.object)
```

Comment:
This feels inefficient but I'm still thinking of a more efficient way to do it.

Reply:
True, I don't love it either but couldn't find another way.